ocaml/typing/typecore.ml

```diff
@@ -475,10 +475,16 @@ let submode ~loc ~env ~reason mode expected_mode =
 let escape ~loc ~env ~reason m =
   submode ~loc ~env ~reason m mode_global
 
-let eqmode ~loc ~env m1 m2 err =
-  match Alloc_mode.equate m1 m2 with
-  | Ok () -> ()
-  | Error () -> raise (Error(loc, env, err))
+let check_eq_mode ~loc ~env m1 cm2 err =
```

Comment:
`eq` is a commutative-sounding name, but `m1` and `cm2` have the same types but don't commute. We should label one or both of them.

Reply:
They do not in fact have the same types anymore: `m1 : Types.Alloc_mode.t` but `cm2 : Types.Alloc_mode.const`.  That said, a better name is still the right call; what do you think about `check_compatible_mode`?

Reply:
Hmm, I don't think "compatible" is much better than "eq". My complaint is maybe more succinctly put as: isn't this checking that `m1 >= cm2`? Is there a way that we can frame this that documents this more plainly? Like are we checking that a value of mode `cm2` can be used at mode `m1`? Then I might suggest `check_value_can_be_used_at_mode : Types.Alloc_mode.const -> at:Types.Alloc_mode.t -> bool`

Reply:
I completely agree with your complaint, just trying to :bike: :derelict_house: the best name.  Global is a submode of local, so the check is more like if `cm2` can be used at `m1`, but it unifies `m1` if necessary.  I think (?) checking against `Global` would be a no-op, so we can skip that step?  So really, this is an implementation of `equate`, in the end.  Maybe `equate_mode_exn`?

Reply:
@riaqn Your proposed change doesn't work: changing this to

```
let check_eq_mode ~loc ~env m1 cm2 err =
  match Alloc_mode.equate m1 (Alloc_mode.of_const cm2) with
  | Ok () -> ()
  | Error () -> raise (Error(loc, env, err))
```

gives errors when compiling the standard library:

```
File "/workspace_root/ocaml/stdlib/camlinternalComprehension.ml", line 1:
Error: The implementation ocaml/stdlib/camlinternalComprehension.ml
       does not match the interface ocaml/stdlib/.stdlib.objs/byte/camlinternalComprehension.cmi:
        Values do not match:
          val rev_list_to_list : 'a rev_list -> 'a list
        is not included in
          val rev_list_to_list : local_ 'a rev_list -> 'a list
        The type 'a rev_list -> 'a list is not compatible with the type
          local_ 'a rev_list -> 'a list
        File "/workspace_root/ocaml/stdlib/camlinternalComprehension.mli", line 23, characters 0-52:
          Expected declaration
        File "/workspace_root/ocaml/stdlib/camlinternalComprehension.ml", line 14, characters 4-20:
          Actual declaration
File "/workspace_root/ocaml/stdlib/iarray.ml", line 512, characters 59-69:
512 | let[@inline always] globalize_bool : local_ bool -> bool = fun b -> b
                                                                 ^^^^^^^^^^
Error: This function has a local parameter, but was expected to have type:
       local_ bool -> bool
File "/workspace_root/ocaml/stdlib/iarrayLabels.ml", line 487, characters 59-69:
487 | let[@inline always] globalize_bool : local_ bool -> bool = fun b -> b
                                                                 ^^^^^^^^^^
Error: This function has a local parameter, but was expected to have type:
       local_ bool -> bool
```

Reply:
Sorry I got it wrong. I think function argument mode annotation should work like this: if I write `local_`, then the function argument must be local. If I don't write anything, then it's not constrained. This matches the logic in the original `typecore.ml` L5819. So I think a better type to represent annotation would be `Value_mode.t option`. This is also how we do it in the [uniqueness PR](https://github.com/ocaml-flambda/ocaml-jst/pull/138/files#diff-154f81eaa30645133fc97f3cd10102b14c29295eff4e36cb8982a33ae884b3d7R722) (See `typecore.ml` L722).

Reply:
ecba032e9 does this and renames the function to `alloc_mode_satisfies_annotation`.  @ncik-roberts, does this put things into a good state?

Reply:
Looks good from my perspective.